### ISSUE TITLE
win: Support ANSI Codes in new Windows 10

### ISSUE THREAD

**Issue by LPGhatguy** [libuv/libuv#721]
*This is copied from chalk/chalk#98; I was directed here.*

It appears that [ANSI color codes have arrived in Windows 10](https://twitter.com/shanselman/status/695738082209853440) in the fast ring insider builds.

Would it be possible to target these new builds with ANSI color and formatting codes? Would it be useful to do that?

[Detection of support should just be a flag](https://twitter.com/shanselman/status/695771659215572993).

**Comment by saghul** [comment]
Since it can be detected at runtime, I guess pull requests are welcome :-)

**Comment by iWuzHere** [comment]
I would like to try and work on this (to get some experience on the windows side). However I cant seem to find much information on this outside of the twitter conversation. Is there some documentation you can point me towards as far as getting started with this goes? (I have already signed up for windows insider builds)

**Comment by okuoku** [comment]
See https://msdn.microsoft.com/en-us/library/mt638032%28v=vs.85%29.aspx for MSDN reference.

I'm not really sure how do we integrate it into libuv though.. As you know, libuv has their own ANSI emulation in `win/tty.c`, so I guess just checking whether or not `ENABLE_VIRTUAL_TERMINAL_PROCESSING`(0x4) enabled (with `GetConsoleMode`) and just disable libuv's emulation according on it is what I could do here...

**Comment by iWuzHere** [comment]
Well just got a win10 vm setup, currently updating it to latest fast ring insider build. Once I have build environment setup I'll try it out and see what happens!

**Comment by saghul** [comment]
@iWuzHere Thanks for volunteering! Alas I won't be able to help you much here...

**Comment by piscisaureus** [comment]
> Would it be possible to target these new builds with ANSI color and formatting codes? Would it be useful to do that?

Possible? sure!
Useful? not sure - libuv already supports ansi escape codes, right?

**Comment by txdv** [comment]
It is super useful! Don't listen to that man, his life is all gray.

**Comment by piscisaureus** [comment]
lol

**Comment by piscisaureus** [comment]
But seriously, @txdv & @iWuzHere.
I'm happy that someone took notice and wants to update libuv to support new features.
However I really don't get it, what functionality does this add/change? please educate me

**Comment by iWuzHere** [comment]
I *think* it's so we dont end up emulating tty's on windows.
I just offered to work on it to get some more windows experience! :stuck_out_tongue:

**Comment by iWuzHere** [comment]
So sorry for the delay (had to clear AIX stuff off my plate) just getting around it to.
Have some example code in https://github.com/iWuzHere/poc-win10ansi

First compiled without the flag ``ENABLE_VIRTUAL_TERMINAL_PROCESSING`` and asserts, ran the code.
Then compiled and ran the code as is without commenting out anything

Got this:
![Screenshot](http://i.imgur.com/PbwNy1r.png)

**Comment by iWuzHere** [comment]
Okay so unfortunately I dont think this will work, or at least I dont know how to make this work ( maybe some ifdef's? )

So in `uv_tty_set_mode` we have the following code

```
switch (mode) {
    case UV_TTY_MODE_NORMAL:
      flags = ENABLE_ECHO_INPUT | ENABLE_LINE_INPUT | ENABLE_PROCESSED_INPUT;
      break;
    case UV_TTY_MODE_RAW:
      flags = ENABLE_WINDOW_INPUT;
      break;
    case UV_TTY_MODE_IO:
      return UV_ENOTSUP;
  }
```

So `ENABLE_ECHO_INPUT` has the same value as `ENABLE_VIRTUAL_TERMINAL_PROCESSING` (0x4)
And coincidentally (i think?) the input maps seem to map to the output flags
https://msdn.microsoft.com/en-us/library/windows/desktop/ms683167(v=vs.85).aspx
 `ENABLE_LINE_INPUT` maps to `ENABLE_WRAP_AT_EOL_OUTPUT`(0x2)
`ENABLE_PROCESSED_INPUT` maps to `ENABLE_PROCESSED_OUTPUT`(0x1)

So the code (sloppy, sorry, in the midst of debugging/learning) I have in `uv_tty_write_bufs` checks to see if the flag is set and just converts to utf16 and writes to console right away and skips the whole conversion thing in that giant for loop

```
if (has_virtual_terminal_processing) {

		//WriteConsoleW(handle->handle, buf.base, buf.len, NULL, NULL);
		puts("in my code");
		utf16_buf_used = MultiByteToWideChar(CP_UTF8, 0, buf.base, buf.len, NULL, 0);
		printf("---BUFFER: %s---", buf.base);
		MultiByteToWideChar(CP_UTF8, 0, buf.base, buf.len, utf16_buf, utf16_buf_used);
		printf("---CONVBUFFER: %ws---", utf16_buf);
		FLUSH_TEXT();
		continue;
	}
```

Now the issue is the flag will be set regardless of windows version if `UV_TTY_MODE_NORMAL` is specified, which means it will skip the whole conversion logic in older windows and thus not  print out our lovely color stuff.

Is there any way to ifdef around MSVER win10preview1511 or something?

So if we had something like that we would now need to support 2 versions of the write code unless MS plans to backport support to all its OS'es in which case we can switch over to the new method.
Otherwise I dont really see the value in this work as the old/current method still works on win10.


TL;DR By happenstance the flag is already being set (0x4) thus any code I will add to skip over the current conversion/emulation will occur on all windows and we will only get ANSI code support in win10 v1511 and newer, unless there is a clever way to ifdef around it.

**Comment by LPGhatguy** [comment]
This feature would also solve #439 and #440 for newer Windows builds, which would be handy.

**Comment by txdv** [comment]
So it is working out of the box?

**Comment by iWuzHere** [comment]
So I thought I could use `GetVersionEx` to try and see if we are using windows 10 v1511 or newer, but its not much help


```
/* OSVERSIONINFO */
Major Version: 6
Minor Version: 2
Build Number: 9200
PlatformID: 2
String:

/* OSVERSIONINFOEX */
Major Version: 6
Minor Version: 2
Build Number: 9200
PlatformID: 2
String:
Service Pack Major: 0
Service Pack Minor: 0
Suite Mask: 0x100
Product Type: 0x1
```

This is for:
![Screenshot](http://i.imgur.com/Z3aPYTV.png)

Looks like windows has gone and made it quite weird to get information version:
https://msdn.microsoft.com/en-us/library/windows/desktop/dn481241(v=vs.85).aspx

**Comment by iWuzHere** [comment]
Managed to get this:

`Windows version: 10.0.14342`
From using `GetFileVersionInfo` and `VerQueryValue` on kernel32.dll

And from [here](https://en.wikipedia.org/wiki/Windows_10_version_history) we find out the build number of 1511 should be `10.0.10586`
Though im not seeing anything about having ansi support in the changelogs

**Comment by iWuzHere** [comment]
Created pull request https://github.com/libuv/libuv/pull/889

**Comment by iWuzHere** [comment]
I guess this issue can be closed now as of #889 
Closing now unless there are any objections

**Comment by saghul** [comment]
Oh, I had forgotten about this one 👍
